cli/open.go

```diff
@@ -0,0 +1,268 @@
+package cli
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/skratchdot/open-golang/open"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/cli/clibase"
+	"github.com/coder/coder/v2/cli/cliui"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+func (r *RootCmd) open() *clibase.Cmd {
+	cmd := &clibase.Cmd{
+		Use:   "open",
+		Short: "Open a workspace",
+		Handler: func(inv *clibase.Invocation) error {
+			return inv.Command.HelpHandler(inv)
+		},
+		Children: []*clibase.Cmd{
+			r.openVSCode(),
+		},
+	}
+	return cmd
+}
+
+const vscodeDesktopName = "VS Code Desktop"
+
+func (r *RootCmd) openVSCode() *clibase.Cmd {
+	var (
+		generateToken bool
+		testOpenError bool
+	)
+
+	client := new(codersdk.Client)
+	cmd := &clibase.Cmd{
+		Annotations: workspaceCommand,
+		Use:         "vscode <workspace> [<directory in workspace>]",
+		Short:       "Open a workspace in Visual Studio Code",
+		Middleware: clibase.Chain(
+			clibase.RequireRangeArgs(1, 2),
+			r.InitClient(client),
+		),
+		Handler: func(inv *clibase.Invocation) error {
+			ctx, cancel := context.WithCancel(inv.Context())
+			defer cancel()
+
+			// Check if we're inside a workspace, and especially inside _this_
+			// workspace so we can perform path resolution/expansion. Generally,
+			// we know that if we're inside a workspace, `open` can't be used.
+			insideAWorkspace := inv.Environ.Get("CODER") == "true"
+			inWorkspaceName := inv.Environ.Get("CODER_WORKSPACE_NAME") + "." + inv.Environ.Get("CODER_WORKSPACE_AGENT_NAME")
+
+			// We need a started workspace to figure out e.g. expanded directory.
+			// Pehraps the vscode-coder extension could handle this by accepting
+			// default_directory=true, then probing the agent. Then we wouldn't
+			// need to wait for the agent to start.
+			workspaceQuery := inv.Args[0]
+			autostart := true
+			workspace, workspaceAgent, err := getWorkspaceAndAgent(ctx, inv, client, autostart, codersdk.Me, workspaceQuery)
+			if err != nil {
+				return xerrors.Errorf("get workspace and agent: %w", err)
+			}
+
+			workspaceName := workspace.Name + "." + workspaceAgent.Name
+			insideThisWorkspace := insideAWorkspace && inWorkspaceName == workspaceName
+
+			if !insideThisWorkspace {
+				// We could optionally add a flag to skip wait, like with SSH.
+				wait := false
+				for _, script := range workspaceAgent.Scripts {
+					if script.StartBlocksLogin {
+						wait = true
+						break
+					}
+				}
```

Comment:
Maybe not. Two reasons I put it here:
1. We don't support waiting in the VS Code extension (yet, e.g. #10473)
2. It's more likely that the agent has reported in once the scripts have executed

That said, it might be better to check for the lifecycle state of the agent, so perhaps I'll just switch to a websocket here if the lifecycle isn't >=starting.